routing: Cancelable payment loop

hieblmi opened the pull request:
Fixes https://github.com/lightningnetwork/lnd/issues/8534.

While interrupting `lncli estimateroutefee` ends the client program the route server is left behind with an abandoned payment loop.

To allow users to cancel payment requests properly the payment loop in https://github.com/lightningnetwork/lnd/blob/f523f52ab468915d9cf6e18632f17ba707906284/routing/payment_lifecycle.go#L170 has to be abandoned as well as soon as the user request it.

This PR makes the `SendPaymentV2` streaming context accessible from the payment loop and allows to safely end it if the context was cancelled.

hieblmi commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -987,7 +989,7 @@ func (s *Server) SetMissionControlConfig(ctx context.Context,
 				),
 				AprioriWeight:    float64(req.Config.Weight),
-				CapacityFraction: routing.DefaultCapacityFraction,
+				CapacityFraction: routing.DefaultCapacityFraction, //nolint:lll
```
move to previous commit

ellemouton commented on routing/payment_lifecycle.go:
```diff
@@ -346,6 +356,33 @@ func (p *paymentLifecycle) checkTimeout() error {
 
+// checkContext checks whether the payment context has been cancelled.
+func (p *paymentLifecycle) checkContext(ctx context.Context) error {
+	if ctx == nil {
+		return nil
+	}
```
I think it is a bit of an anti-pattern to ever allow passing in a nil context. I think we should just create one at the call-sites where one is not available

ellemouton commented on routing/payment_lifecycle.go:
```diff
@@ -346,6 +356,33 @@ func (p *paymentLifecycle) checkTimeout() error {
 }
 
+// checkContext checks whether the payment context has been cancelled.
+func (p *paymentLifecycle) checkContext(ctx context.Context) error {
```
since both this and `checkTimeout` do basically the same thing, and since they are only called from one spot and since I think we would always want to take the same failure steps no matter which one (context or timeout) causes the trigger, perhaps we can combine them instead to avoid the duplicate logic?

```
func (p *paymentLifecycle) checkTimeoutAndContext(ctx context.Context) error {
	failPayment := func(reason channeldb.FailureReason) error {
		// By marking the payment failed, depending on whether it has
		// inflight HTLCs or not, its status will now either be
		// `StatusInflight` or `StatusFailed`. In either case, no more
		// HTLCs will be attempted.
		err := p.router.cfg.Control.FailPayment(p.identifier, reason)
		if err != nil {
			return fmt.Errorf("FailPayment got %w", err)
		}

		return nil
	}

	select {
	case <-p.timeoutChan:
		log.Warnf("payment attempt not completed before timeout")
		err := failPayment(channeldb.FailureReasonTimeout)
		if err != nil {
			return err
		}

	case <-ctx.Done():
		log.Warnf("payment attempt context canceled")
		err := failPayment(channeldb.FailureReasonError)
		if err != nil {
			return err
		}

		return ctx.Err()

	case <-p.router.quit:
		return fmt.Errorf("check payment timeout got: %w",
			ErrRouterShuttingDown)

	// Fall through if we haven't hit our time limit.
	default:
	}

	return nil
}
```

ellemouton replied:
non blocking at all btw. just a suggestion - very open to being wrong here :)

hieblmi replied:
I first had the context check included in `checkTimeout` but then thought that the separation would make it clearer that we are adding an extra step in the payment status evaluation. But I agree the code dupe is not nice, so I also think its cleaner to merge the two checks.

ellemouton reviewed (commented):
Looking good!

Only one blocking comment around removing the `context.Context` nil check.

The other comment is just a suggestion

ellemouton reviewed (approved):
LGTM! 🚀

yyforyongyu reviewed (commented):
Don't we have timeout already? Why do we need this? I think once the payment is sent, there's no easy way to cancel it, maybe related #5461

hieblmi commented:
That's also my understanding, we can't cancel the payment once the htlc is sent. Here, additionally to the existent timeout check that interrupts the payment loop on expiry, we pass in a context (from the rpc server client stream) that is checked for cancellation and interrupts the payment loop if the user cancelled the stream context(Ctrl+C'd the sendpayment from cli). This can only happen if a `p.sendAttempt(attempt)` completed.

bitromortac commented on routing/payment_lifecycle_test.go:
```diff
@@ -727,6 +762,49 @@ func TestResumePaymentFailOnTimeoutErr(t *testing.T) {
+	cancel()
+
+	// Send the payment and assert that its context got cancelled.
+	sendPaymentAndAssertContextCancelled(t, ctx, p, ErrContextCancelled)
```
nit: `ErrContextCancelled` isn't used elsewhere, you could use `context.Canceled`.

bitromortac commented on routing/router.go:
```diff
@@ -92,6 +93,10 @@ var (
 
+	// ErrContextCancelled is returned if the context of the payment attempt
+	// is canceled.
+	ErrContextCancelled = errors.New("context cancelled")
```
nit: we can delete this and use `context.Canceled` in tests

bitromortac reviewed (commented):
Code wise this looks good :fire:! Wanted to note that this code may break usage patterns, if users expect to be able to dispatch payments in the background with `lncli sendpayment` or somehow if `SendPaymentV2` is not waited on for responses. This would not be possible anymore with the current setting as it stops the payment loop in the next occasion if the context is cancelled. If the cmd is interrupted, this will lead to cancellation of the payment loop, as expected. Should we retain backward compatibility by adding a `bool` `cancellable` to the payment request (we could then decide whether we'd like to forward the context or to initialize a new background one)?

bhandras commented on routing/payment_lifecycle.go:
```diff
@@ -319,27 +322,42 @@ lifecycle:
+			return err
+		}
+
+	case <-ctx.Done():
```
I'd argue, that if the context times out we could instead fail the payment with `FailureReasonTimeout`. Would need to check if the `ctx.Err()` is `context.DeadlineExceeded`.

hieblmi replied:
I'd like to get rid of `p.timeoutChan` by moving the timeout to the context (but only if `cancelable` is set on the payRequest - see @bitromortac's comment), however then we couldn't distinguish between a context cancellation and a timeout cancellation inside the payment loop.

hieblmi replied:
Thanks for pointing this out. I've removed the `p.timeoutChan` and replaced it with a deadline dependent context.

yyforyongyu commented on routing/payment_lifecycle.go:
```diff
@@ -319,27 +322,42 @@ lifecycle:
+			return err
+		}
+
+		return ctx.Err()
```
Unfortunately we cannot return an error here, as it'd cause the payment loop to exit. For the timeout case, we never exit the loop. Consider this case, say a payment is made of two HTLC attempts,
```
HTLC1 ---> hop1 ---> remote
HTLC2 ---> hop2 ---> remote
```
Then remote settles the invoice, but hop2 is offline, we'd end up in this case,
```
HTLC1 ---> hop1 ---> remote ---> settled
HTLC2 ---> hop2 ---> remote ---> pending
```
After 60s, we'd time out this payment, but we won't quit the loop as we still think the payment is inflight,
https://github.com/lightningnetwork/lnd/blob/9d358bc64934caef724108fc9704097db31658c4/channeldb/payment_status.go#L148

Then hop2 comes online and settles it, we'd consider the payment successful.
https://github.com/lightningnetwork/lnd/blob/9d358bc64934caef724108fc9704097db31658c4/channeldb/payment_status.go#L156

yyforyongyu replied:
In other words, once the HTLC is sent, there's no easy way to cancel it unless you get a clear response from your peer saying it's failed/settled. If we exit here because the context is canceled, the payment would still have the status `StatusInFlight`.

hieblmi replied:
Thank you for the detailed explanation. I see why we can't interrupt the entire payment loop but instead behave exactly how we do it the timeout case.

hieblmi commented on routing/payment_lifecycle.go:
```diff
@@ -319,27 +322,42 @@ lifecycle:
+	select {
+	case <-p.timeoutChan:
+		log.Warnf("payment attempt not completed before timeout")
+		err := failPayment(channeldb.FailureReasonTimeout)
```
good catch

hieblmi commented on routing/payment_lifecycle_test.go:
```diff
@@ -179,6 +180,34 @@ func sendPaymentAndAssertFailed(t *testing.T,
 
+// sendPaymentAndAssertFailed calls `resumePayment` and asserts that an error
+// is returned.
+func sendPaymentAndAssertContextCancelled(t *testing.T,
```
I've consolidated the two.

hieblmi commented on routing/payment_lifecycle_test.go:
```diff
@@ -179,6 +180,34 @@ func sendPaymentAndAssertFailed(t *testing.T,
 	}
 }
 
+// sendPaymentAndAssertFailed calls `resumePayment` and asserts that an error
```
I've called it `sendPaymentAndAssertError`.

hieblmi commented:
Thanks to the reviewers so far. I think I addressed all your concerns, namely:

- Add a `cacelable` flag to the `SendPaymentRequest` that preserves the current behavior of a long running payment loop even if the stream context is canceled, but cancels the payment loop if set to true.
- Wrap the send payment context into a deadline context in case the a timeout is set in the `SendPaymentRequest`.
- Remove the timeout channel in from the `payment_lifecycle.go` in favor of the new timeout context.

If tested different scenarios locally and didn't encounter an error, additionally to the added test.

Looking forward to another round of feedback.

bitromortac commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -360,13 +360,37 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
+
+	// Cancel the timeout context. If the context already timed out or if
+	// there was no timeout provided, this will be a no-op.
+	cancel()
```
this doesn't cancel the payment directly because `trackPayment` is blocking?

hieblmi replied:
This call is to avoid a linter issue that would arise if I wrote above `ctx, _  = context.WithDeadline(ctx, timeout)`. And if I assign the cancel I have to call it. The right place seems to be after `trackPayment` because then the context can be canceled.

bitromortac replied:
an alternative could be to place it after `sendPayment` in `SendPaymentAsync`, but not sure that's a good pattern

hieblmi replied:
could also silence the linter here with a comment as to why we don't need the context in this case.

hieblmi replied:
Maybe @yyforyongyu could opine here as this concerns his suggestion to remove the timeout channel in favor of this context?

yyforyongyu replied:
If we call cancel here I think it will immediately invoke `<-ctx.Done` in `checkContext` since it's async sending. I think we can instead decide the right context to use here (`stream.Context` or background ctx), pass it to `SendPaymentAsync -> sendPayment`, and inside `sendPayment` we create the timeout ctx, and call the `defer cancel()` there.

yyforyongyu replied:
This timeout thing may also be moved to the payment session, but that's a bigger refactor so future PRs.

hieblmi replied:
> If we call cancel here I think it will immediately invoke `<-ctx.Done` in `checkContext` since it's async sending.

Before `cancel()` is called there is the `s.trackPayment` call which is blocking. So I think `cancel()` is only called once the track payment call ends which implies that also the async payment loop ended. I think the cancel would then be correct here. We could add clarifying docs here. 

Could also create the timeoutCtx in `sendPayment`. instead.

yyforyongyu replied:
> Before cancel() is called there is the s.trackPayment call which is blocking.

Gotya - yea to me it feels "safer" to do this in `sendPayment`, so any future changes in `s.trackPayment` won't affect the behavior of sending payments.

hieblmi replied:
Okay I will change this.

bitromortac commented on routing/payment_lifecycle.go:
```diff
@@ -318,23 +312,45 @@ lifecycle:
+// checkContext checks whether the payment context has been canceled.
+// Cancellation occurs manually or if the context times out.
+func (p *paymentLifecycle) checkContext(ctx context.Context) error {
+	failPayment := func(reason channeldb.FailureReason) error {
```
I think this can be inlined again

bitromortac commented on routing/payment_lifecycle.go:
```diff
@@ -318,23 +312,45 @@ lifecycle:
+	case <-ctx.Done():
+		failureReason := channeldb.FailureReasonError
+
+		switch {
```
This could be an if/else instead

bitromortac reviewed (approved):
tACK LGTM :zap:. Perhaps the error could be changed to a timeout error and there's an open question concerning the `cancel` call.

yyforyongyu commented on routing/payment_lifecycle.go:
```diff
@@ -318,23 +312,36 @@ lifecycle:
-		err := p.router.cfg.Control.FailPayment(
-			p.identifier, channeldb.FailureReasonTimeout,
-		)
+		reason := channeldb.FailureReasonTimeout
```
👍

yyforyongyu commented on routing/payment_lifecycle.go:
```diff
@@ -318,23 +312,36 @@ lifecycle:
 			return fmt.Errorf("FailPayment got %w", err)
 		}
 
+		return nil
```
nit: don't think we need to return here

yyforyongyu commented on routing/payment_lifecycle.go:
```diff
@@ -318,23 +312,36 @@ lifecycle:
 
-// checkTimeout checks whether the payment has reached its timeout.
-func (p *paymentLifecycle) checkTimeout() error {
+// checkContext checks whether the payment context has been canceled.
```
or timed out...

hieblmi replied:
The second sentence clarifies that.

yyforyongyu commented on routing/router.go:
```diff
@@ -2698,8 +2712,8 @@ func (r *ChannelRouter) sendToRoute(htlcHash lntypes.Hash, rt *route.Route,
-func (r *ChannelRouter) sendPayment(feeLimit lnwire.MilliSatoshi,
-	identifier lntypes.Hash, timeout time.Duration,
+func (r *ChannelRouter) sendPayment(ctx context.Context,
+	feeLimit lnwire.MilliSatoshi, identifier lntypes.Hash,
```
so we could pass the timeout value here, and create a timeout context from the parent context.

hieblmi replied:
Done.

yyforyongyu commented on routing/router.go:
```diff
@@ -2699,11 +2698,23 @@ func (r *ChannelRouter) sendToRoute(htlcHash lntypes.Hash, rt *route.Route,
+	cancel := func() {}
+	if paymentAttemptTimeout > 0 {
+		timeout := time.Now().Add(paymentAttemptTimeout)
+		ctx, cancel = context.WithDeadline(ctx, timeout)
```
we can call `defer cancel()` here to avoid declaring it?

hieblmi replied:
I replaced the `WithDeadline` with `WithTimeout`, but I still have to declare `cancel` since I can't re-declare `ctx` here.

yyforyongyu replied:
oh that's right, yea in this case it's fine as-is.

yyforyongyu commented on docs/release-notes/release-notes-0.18.1.md:
```diff
@@ -19,14 +19,23 @@
 
 # Bug Fixes
 
+* [SendPaymentV2]() now cancels the background payment loop if the user cancels
```
missing the PR link

yyforyongyu commented on routing/payment_lifecycle.go:
```diff
@@ -318,19 +312,30 @@ lifecycle:
+				"timeout, id=%s", p.identifier.String())
+		} else {
+			log.Warnf("Payment attempt context canceled, id=%s",
+				p.identifier.String())
```
Let's add a new failure `FailureReasonCanceled` to record the failure reason explicitly - think it will save us dome debugging timeout in the future. Also prefer doing it in a commit, or better a new PR since this PR is already very complete.

hieblmi replied:
Good call, I addressed this in this follow-up PR: https://github.com/lightningnetwork/lnd/pull/8836

yyforyongyu reviewed (commented):
Left a few comments, and needs a rebase!

yyforyongyu commented:
Just merged the PR that fixed the itest, could you do a rebase and let it run again, thanks!

hieblmi commented:
There still seems to be an unrelated error left: https://github.com/lightningnetwork/lnd/actions/runs/9501738226/job/26188011967?pr=8734#step:6:468

yyforyongyu reviewed (approved):
Yea that test is a flake, and thanks for the PR! LGTM☄️
